File: src/main/java/com/menmar/gestionTienda/service/CatalogoService.java
package com.menmar.gestionTienda.service;

import com.menmar.gestionTienda.model.catalogo.*;

import java.util.List;

/**
 * Gestión del catálogo de la tienda: familias de reparación,
 * tipos de reparación de calzado, costuras y llaves.
 * <p>
 * Las operaciones de lectura están disponibles para todos los roles.
 * Las de escritura requieren {@code ADMIN}.
 */
public interface CatalogoService {

    // ---- Familias ----

    /** Devuelve todas las familias de reparación activas. */
    List<FamiliaReparacionResponse> listarFamilias();

    /**
     * Busca una familia por su ID.
     *
     * @throws jakarta.persistence.EntityNotFoundException si no existe
     */
    FamiliaReparacionResponse buscarFamilia(Long id);

    /** Crea una nueva familia de reparación. */
    FamiliaReparacionResponse crearFamilia(FamiliaReparacionRequest request);

    /** Actualiza el nombre de una familia existente. */
    FamiliaReparacionResponse actualizarFamilia(Long id, FamiliaReparacionRequest request);

    /** Desactiva (soft delete) una familia. */
    void desactivarFamilia(Long id);

    // ---- Tipos reparación calzado ----

    /**
     * Lista tipos de reparación de calzado.
     * Si {@code familiaId} no es nulo, filtra por familia.
     */
    List<TipoReparacionCalzadoResponse> listarTiposCalzado(Long familiaId);

    /** @throws jakarta.persistence.EntityNotFoundException si no existe */
    TipoReparacionCalzadoResponse buscarTipoCalzado(Long id);

    TipoReparacionCalzadoResponse crearTipoCalzado(TipoReparacionCalzadoRequest request);

    TipoReparacionCalzadoResponse actualizarTipoCalzado(Long id, TipoReparacionCalzadoRequest request);

    void desactivarTipoCalzado(Long id);

    // ---- Tipos costura ----

    List<TipoCosturaResponse> listarTiposCostura();

    /** @throws jakarta.persistence.EntityNotFoundException si no existe */
    TipoCosturaResponse buscarTipoCostura(Long id);

    TipoCosturaResponse crearTipoCostura(TipoCosturaRequest request);

    TipoCosturaResponse actualizarTipoCostura(Long id, TipoCosturaRequest request);

    void desactivarTipoCostura(Long id);

    // ---- Tipos llave ----

    List<TipoLlaveResponse> listarTiposLlave();

    /** @throws jakarta.persistence.EntityNotFoundException si no existe */
    TipoLlaveResponse buscarTipoLlave(Long id);

    TipoLlaveResponse crearTipoLlave(TipoLlaveRequest request);

    TipoLlaveResponse actualizarTipoLlave(Long id, TipoLlaveRequest request);

    void desactivarTipoLlave(Long id);
}

File: frontend/src/service/CatalogoService.ts
import api from './ApiService'
import type {
    FamiliaReparacionRequest,
    FamiliaReparacionResponse,
    TipoCosturaRequest,
    TipoCosturaResponse,
    TipoLlaveRequest,
    TipoLlaveResponse,
    TipoReparacionCalzadoRequest,
    TipoReparacionCalzadoResponse
} from '@/types'

const CatalogoService = {
    // Familias
    async listarFamilias(): Promise<FamiliaReparacionResponse[]> {
        const res = await api.get<FamiliaReparacionResponse[]>('/catalogo/familias')
        return res.data
    },
    async crearFamilia(req: FamiliaReparacionRequest): Promise<FamiliaReparacionResponse> {
        const res = await api.post<FamiliaReparacionResponse>('/catalogo/familias', req)
        return res.data
    },
    async actualizarFamilia(id: number, req: FamiliaReparacionRequest): Promise<FamiliaReparacionResponse> {
        const res = await api.put<FamiliaReparacionResponse>(`/catalogo/familias/${id}`, req)
        return res.data
    },
    async eliminarFamilia(id: number): Promise<void> {
        await api.delete(`/catalogo/familias/${id}`)
    },

    // Calzado
    async listarCalzado(familiaId?: number): Promise<TipoReparacionCalzadoResponse[]> {
        const res = await api.get<TipoReparacionCalzadoResponse[]>('/catalogo/reparaciones-calzado', {
            params: { familiaId }
        })
        return res.data
    },
    async crearCalzado(req: TipoReparacionCalzadoRequest): Promise<TipoReparacionCalzadoResponse> {
        const res = await api.post<TipoReparacionCalzadoResponse>('/catalogo/reparaciones-calzado', req)
        return res.data
    },
    async actualizarCalzado(id: number, req: TipoReparacionCalzadoRequest): Promise<TipoReparacionCalzadoResponse> {
        const res = await api.put<TipoReparacionCalzadoResponse>(`/catalogo/reparaciones-calzado/${id}`, req)
        return res.data
    },
    async eliminarCalzado(id: number): Promise<void> {
        await api.delete(`/catalogo/reparaciones-calzado/${id}`)
    },

    // Costura
    async listarCosturas(): Promise<TipoCosturaResponse[]> {
        const res = await api.get<TipoCosturaResponse[]>('/catalogo/costuras')
        return res.data
    },
    async crearCostura(req: TipoCosturaRequest): Promise<TipoCosturaResponse> {
        const res = await api.post<TipoCosturaResponse>('/catalogo/costuras', req)
        return res.data
    },
    async actualizarCostura(id: number, req: TipoCosturaRequest): Promise<TipoCosturaResponse> {
        const res = await api.put<TipoCosturaResponse>(`/catalogo/costuras/${id}`, req)
        return res.data
    },
    async eliminarCostura(id: number): Promise<void> {
        await api.delete(`/catalogo/costuras/${id}`)
    },

    // Llaves
    async listarLlaves(): Promise<TipoLlaveResponse[]> {
        const res = await api.get<TipoLlaveResponse[]>('/catalogo/llaves')
        return res.data
    },
    async crearLlave(req: TipoLlaveRequest): Promise<TipoLlaveResponse> {
        const res = await api.post<TipoLlaveResponse>('/catalogo/llaves', req)
        return res.data
    },
    async actualizarLlave(id: number, req: TipoLlaveRequest): Promise<TipoLlaveResponse> {
        const res = await api.put<TipoLlaveResponse>(`/catalogo/llaves/${id}`, req)
        return res.data
    },
    async eliminarLlave(id: number): Promise<void> {
        await api.delete(`/catalogo/llaves/${id}`)
    }
}

export default CatalogoService
